Remove fixed_bytes macro

bewakes opened the pull request:
## Description
Fixes: https://github.com/alpenlabs/bitcoin-bosd/issues/32

This PR removes the `fixed_bytes!` macro, and equivalent method is added to the `Descriptor` implementation.

### Type of Change

<!--
Select the type of change your PR introduces (put an `x` in all that apply):
-->

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature/Enhancement (non-breaking change which adds functionality or enhances an existing one)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Documentation update
- [ ] Refactor
- [ ] New or updated tests
- [ ] Dependency Update

## Notes to Reviewers

<!--
Anything in particular you want to note that will help reviewers fulfill their role
in reviewing this PR?
-->

## Checklist

<!--
Ensure all the following are checked:
-->

- [x] I have performed a self-review of my code.
- [x] I have commented my code where necessary.
- [ ] I have updated the documentation if needed.
- [x] My changes do not introduce new warnings.
- [x] I have added tests that prove my changes are effective or that my feature works.
- [x] New and existing tests pass with my changes.

## Related Issues

<!--
Link any related issues (e.g., `closes #123`, `fixes #456`).
-->

AaronFeickert commented on src/address.rs:
```diff
@@ -21,39 +21,34 @@ impl Descriptor {
             P2pkh => {
-                fixed_bytes!(20);
-                let bytes = to_fixed_bytes(self);
+                let bytes = self.to_fixed_bytes::<20>();
```
Can this magic number be replaced by a descriptor constant?

AaronFeickert commented on src/address.rs:
```diff
@@ -21,39 +21,34 @@ impl Descriptor {
             P2sh => {
-                fixed_bytes!(20);
-                let bytes = to_fixed_bytes(self);
+                let bytes = self.to_fixed_bytes::<20>();
```
Can this magic number be replaced by a descriptor constant?

AaronFeickert commented on src/address.rs:
```diff
@@ -21,39 +21,34 @@ impl Descriptor {
             P2wpkh => {
-                fixed_bytes!(20);
-                let bytes = to_fixed_bytes(self);
+                let bytes = self.to_fixed_bytes::<20>();
```
Can this magic number be replaced by a descriptor constant?

AaronFeickert commented on src/address.rs:
```diff
@@ -21,39 +21,34 @@ impl Descriptor {
             P2wsh => {
-                fixed_bytes!(32);
-                let bytes = to_fixed_bytes(self);
+                let bytes = self.to_fixed_bytes::<32>();
```
Can this magic number be replaced by a descriptor constant?

AaronFeickert commented on src/address.rs:
```diff
@@ -21,39 +21,34 @@ impl Descriptor {
             P2tr => {
-                fixed_bytes!(32);
-                let bytes = to_fixed_bytes(self);
+                let bytes = self.to_fixed_bytes::<32>();
```
Can this magic number be replaced by a descriptor constant?

AaronFeickert commented on src/address.rs:
```diff
@@ -95,36 +90,31 @@ impl Descriptor {
             P2pkh => {
-                fixed_bytes!(20);
-                let bytes = to_fixed_bytes(self);
+                let bytes = self.to_fixed_bytes::<20>();
```
Can this magic number be replaced by a descriptor constant?

AaronFeickert commented on src/address.rs:
```diff
@@ -95,36 +90,31 @@ impl Descriptor {
             P2sh => {
-                fixed_bytes!(20);
-                let bytes = to_fixed_bytes(self);
+                let bytes = self.to_fixed_bytes::<20>();
```
Can this magic number be replaced by a descriptor constant?

AaronFeickert commented on src/address.rs:
```diff
@@ -95,36 +90,31 @@ impl Descriptor {
             P2wpkh => {
-                fixed_bytes!(20);
-                let bytes = to_fixed_bytes(self);
+                let bytes = self.to_fixed_bytes::<20>();
```
Can this magic number be replaced by a descriptor constant?

AaronFeickert commented on src/address.rs:
```diff
@@ -95,36 +90,31 @@ impl Descriptor {
             P2wsh => {
-                fixed_bytes!(32);
-                let bytes = to_fixed_bytes(self);
+                let bytes = self.to_fixed_bytes::<32>();
```
Can this magic number be replaced by a descriptor constant?

AaronFeickert commented on src/address.rs:
```diff
@@ -95,36 +90,31 @@ impl Descriptor {
             P2tr => {
-                fixed_bytes!(32);
-                let bytes = to_fixed_bytes(self);
+                let bytes = self.to_fixed_bytes::<32>();
```
Can this magic number be replaced by a descriptor constant?

AaronFeickert commented on src/descriptor.rs:
```diff
@@ -479,20 +468,18 @@ mod tests {
         )
         .unwrap();
-        let bytes = to_fixed_bytes(&desc);
+        let bytes = desc.to_fixed_bytes::<32>();
```
Can this magic number be replaced by a descriptor constant?

AaronFeickert commented on src/descriptor.rs:
```diff
@@ -479,20 +468,18 @@ mod tests {
-        fixed_bytes!(20); // Generates to_fixed_bytes for P2PKH/P2SH
         let desc = Descriptor::from_str("01b8268ce4d481413c4e848ff353cd16104291c45b").unwrap();
-        let bytes = to_fixed_bytes(&desc);
+        let bytes = desc.to_fixed_bytes::<20>();
```
Can this magic number be replaced by a descriptor constant?

storopoli reviewed (changes requested):
I am on board with the changes but maybe change the hardcoded numbers for the constants that were already have and are `pub(crate)` as @AaronFeickert suggested.

bewakes commented:
@storopoli @AaronFeickert done in 7ce9f75

storopoli commented on src/descriptor.rs:
````diff
@@ -165,6 +165,43 @@ impl Descriptor {
+    /// assert_eq!(bytes.len(), 32);
+    /// # }
+    /// ```
+    pub fn to_fixed_bytes<const B: usize>(&self) -> [u8; B] {
````
I don't think this function should be `pub` it should be only used internally. But I don't know. Would the user get a compile-time error or a runtime error if used with the wrong `B`?

AaronFeickert replied:
Agreed that it should not be public. And, of course, its use should be covered in unit tests.

AaronFeickert replied:
@storopoli: This is likely a task for a future PR, but I wonder if a better approach is to make the relevant length constants public instead. Then a user who wants to do their own byte-level handling could do so with more information at hand. This would also likely make testing of calling code easier.

storopoli replied:
Yes ACK on the `pub` for the constants.

About the `fixed_bites!` being public was my lack of skills to use it from `descriptor.rs` to `address.rs`...

storopoli replied:
FYI: #34.

bewakes replied:
> I don't think this function should be pub it should be only used internally

I don't see why it can't be public, seems to me pretty usable from outside as well. Plus, it has similar counterpart `to_bytes` public due to which also it kind of makes sense to have it public.


> Would the user get a compile-time error or a runtime error if used with the wrong B?

User would get a compile-time error if the expected B and actual B used differ. For example This would be give a compile-time error:

```rust
let x: [u8; 20] = desc.to_fixed_bytes<32>();
// mismatched types, expected an array with a size of 20, found 32
```

storopoli replied:
Ok, that is good. I also see a valid use case for when you want to discard the first byte (the type tag) and convert the underlying bytes (the payload) into a `[W]ScriptHash`, `XOnlyPublicKey`, `[W]PubkeyHash`, and these methods have easy conversions from `bitcoin_hashes::foo::Hash` that takes an `[u8; N]`, i.e. fixed-size array.

@bewakes, let's keep public but there's one issue here.
Only the `payload` gets to be converted `to_fixed_bytes`.
Whereas the method is on the `Descriptor`.
Now this can be solved two ways:

1. `Descriptor::to_fixed_bytes` returns a byte array with the type tag + payload. Example, P2TR returns a 33-byte array and internally we drop the first byte when doing the conversions.
2. Rename from `Descriptor::to_fixed_bytes` to `Descriptor::payload_to_fixed_bytes` and keep as it is.

I prefer 1. What you think?

bewakes replied:
Done in [de50469](https://github.com/alpenlabs/bitcoin-bosd/pull/31/commits/de50469129cdbf0d6c377a564327ede2438f2f64)

AaronFeickert replied:
Agreed that we should only return the payload in this case. Forcing the caller to do parsing seems like an unnecessary footgun.

AaronFeickert replied:
@bewakes: In `to_fixed_payload_bytes`, it could be the case that there isn't a compile-time error for a particular `B`, but that this value does match the corresponding constant for the underlying type. In this case, one of two things can happen:
- If `B < payload.len()`, only a truncated payload is returned
- If `B > payload.len()`, the payload is padded with zeros

I think neither of these cases is desirable. But @storopoli: is the latter case what you meant above when discussing other conversions requiring fixed-length arrays?

storopoli replied:
Damn... You are right. Let's keep this private for now? And exposed if there are angry mobs with torches asking for it?

@bewakes the original macro was public but not intended to be. I wasn't able to use it from `descriptor.rs` into `address.rs`.

AaronFeickert replied:
Couldn't you also expose the same functionality but return a vector? You pay for it with an allocation, but don't need to worry about the length parameter.

storopoli replied:
![](https://i.imgflip.com/9gjzpt.jpg)

bewakes commented:
Rebased with main, leaving the method public as it is.

storopoli commented:
Also what should it be called?

- `payload_to_fixed_bytes` (my suggestion)
- `to_fixed_payload_bytes` (how it is now)

storopoli commented on src/descriptor.rs:
```diff
@@ -165,6 +165,43 @@ impl Descriptor {
+    /// # use bitcoin_bosd::Descriptor;
+    /// #
+    /// # fn main() {
+    /// let desc = Descriptor::from_str(
+    ///     "040f0c8db753acbd17343a39c2f3f4e35e4be6da749f9e35137ab220e7b238a667",
+    /// ).unwrap();
+    /// let bytes = desc.to_fixed_payload_bytes::<32>();
+    /// assert_eq!(bytes.len(), 32);
```
I don't think this is valid 😂. It was a draft of the intention...

bewakes replied:
Yes, realized it 😆

bewakes commented:
@storopoli , that's a great suggestion. But I think that `to_payload_bytes` is clearer and follows rust conventions(like `to_bytes`). Added the hidden assertion check in [d753bc9](https://github.com/alpenlabs/bitcoin-bosd/commit/d753bc9c0442e87af61d25de4f4a70a041af39a1)

storopoli reviewed (approved):
ACK d753bc9c0442e87af61d25de4f4a70a041af39a1

AaronFeickert commented on src/descriptor.rs:
````diff
@@ -165,6 +165,44 @@ impl Descriptor {
+    /// # assert_eq!(bytes.as_slice(), &Vec::from_hex(raw).unwrap()[1..]);
+    /// # }
+    /// ```
+    pub fn to_fixed_payload_bytes<const B: usize>(&self) -> [u8; B] {
````
As discussed, recommend that this not be public.

AaronFeickert commented on src/descriptor.rs:
````diff
@@ -165,6 +165,44 @@ impl Descriptor {
+    /// ```
+    pub fn to_fixed_payload_bytes<const B: usize>(&self) -> [u8; B] {
+        let mut bytes = [0u8; B];
+        let len = self.payload().len().min(B);
````
This should be replaced with a debug assert of equality.

bewakes commented:
@AaronFeickert @storopoli , I've made the method private and updated the corresponding docs: basically removed the code part. Also added the debut assertations.

Done in [a54b835](https://github.com/alpenlabs/bitcoin-bosd/pull/31/commits/a54b8356cb178e012dc0f3082fbb41b09d5dda49)

storopoli commented on src/descriptor.rs:
```diff
@@ -165,6 +165,18 @@ impl Descriptor {
         bytes
     }
 
+    /// Generates fixed bytes of payload of length specified by the generic parameter.
```
```suggestion
    /// Generates fixed bytes of payload of length specified by the generic parameter.
    ///
```
empty line before header

bewakes replied:
done in [036ce0e](https://github.com/alpenlabs/bitcoin-bosd/pull/31/commits/036ce0e369e4c3ed03578b1f770bff709468d490)

storopoli reviewed (changes requested):
minor nit

storopoli reviewed (approved):
ACK 036ce0e369e4c3ed03578b1f770bff709468d490

storopoli commented:
@AaronFeickert LGTM but I'll wait for your review.

AaronFeickert reviewed (approved):
LGTM
